### PR TITLE
Fix 300s timeout for grpc streams

### PR DESCRIPTION
**Pull request by rshriram**
Disable idle timeouts in HTTP connection manager that causes idle gRPC streams to be terminated every 5 minutes. Along the way, also updated some deprecated fields.

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

File: pilot/pkg/model/service.go [hunk 1]
```diff
@@ -74,10 +74,6 @@ type Service struct {
 	// These services are defined using Istio's ServiceEntry spec.
 	MeshExternal bool
 
-	// LoadBalancingDisabled indicates that no load balancing should be done for this service.
-	// Deprecated : made obsolete by the MeshExternal and Resolution flags.
-	LoadBalancingDisabled bool `json:"-"`
```
Comment (ijsnellf) [inline pilot/pkg/model/service.go]: :+1:

**Comment by costinm** [comment]
seems to fail pretty badly - does it depend on a new envoy ?

**Comment by rshriram** [comment]
@ijsnellf or @GregHanson could you please help with the PR?

**Comment by rshriram** [comment]
All prow errors: “error	Failed to complete Init. Error istio-ingressgateway readiness check timed out”

**Comment by costinm** [comment]
/test 
(merged Envoy)

**Comment by costinm** [comment]
Tried to merge head so circle has the last version of envoy.

**Comment by costinm** [comment]
From the failed simple test:

Pilot logs - shows a disconnect immediately after the RDS push:

```
2018-07-28T00:13:58.976529Z	info	ads	CDS: PUSH for istio-egressgateway-777f778c89-sgdxh.istio-system "127.0.0.1:53508", Clusters: 40
2018-07-28T00:13:59.120760Z	info	ads	LDS: PUSH for node:istio-egressgateway-777f778c89-sgdxh.istio-system addr:"127.0.0.1:53508" listeners:0
2018-07-28T00:14:13.974945Z	info	ads	ADS:CDS: REQ router~172.17.0.7~istio-ingressgateway-68f6748d46-gfcnb.istio-system~istio-system.svc.cluster.local-4 127.0.0.1:55018 raw: node:<id:"router~172.17.0.7~istio-ingressgateway-68f6748d46-gfcnb.istio-system~istio-system.svc.cluster.local" cluster:"istio-ingressgateway" metadata:<fields:<key:"ISTIO_PROXY_SHA" value:<string_value:"istio-proxy:efb5224d4b2249ff3894d989ff7355bb62d8af3e" > > fields:<key:"ISTIO_PROXY_VERSION" value:<string_value:"1.0.0" > > fields:<key:"ISTIO_VERSION" value:<string_value:"1.0-dev" > > fields:<key:"POD_NAME" value:<string_value:"istio-ingressgateway-68f6748d46-gfcnb" > > fields:<key:"istio" value:<string_value:"sidecar" > > > build_version:"0/1.8.0-dev//RELEASE" > type_url:"type.googleapis.com/envoy.api.v2.Cluster"  
2018-07-28T00:14:14.014747Z	info	ads	CDS: PUSH for istio-ingressgateway-68f6748d46-gfcnb.istio-system "127.0.0.1:55018", Clusters: 40
2018-07-28T00:14:14.098806Z	info	ads	LDS: PUSH for node:istio-ingressgateway-68f6748d46-gfcnb.istio-system addr:"127.0.0.1:55018" listeners:1
2018-07-28T00:14:14.180048Z	info	ads	ADS: RDS: PUSH for node: istio-ingressgateway-68f6748d46-gfcnb.istio-system addr:127.0.0.1:55018 routes:1
2018-07-28T00:14:27.978374Z	info	ads	ADS: "127.0.0.1:55018" sidecar~172.17.0.5~echosrv-deployment-2-799cd8d8bd-2sdfp.istio-system~istio-system.svc.cluster.local-2 terminated rpc error: code = Canceled desc = context canceled
2018-07-28T00:14:27.978482Z	info	ads	ADS: "127.0.0.1:55018" router~172.17.0.7~istio-ingressgateway-68f6748d46-gfcnb.istio-system~istio-system.svc.cluster.local-4 terminated rpc error: code = Canceled desc = context canceled
```

Envoy logs for ingress:

```
[2018-07-28 00:14:14.090][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:388] add/update cluster outbound|9102||istio-statsd-prom-bridge.istio-system.svc.cluster.local during init
[2018-07-28 00:14:14.092][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:388] add/update cluster outbound|9125||istio-statsd-prom-bridge.istio-system.svc.cluster.local during init
[2018-07-28 00:14:14.092][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:388] add/update cluster BlackHoleCluster during init
[2018-07-28 00:14:14.092][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:110] cm init: initializing secondary clusters
[2018-07-28 00:14:14.095][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:133] cm init: all clusters initialized
[2018-07-28 00:14:14.095][51][info][main] external/envoy/source/server/server.cc:390] all clusters initialized. initializing init manager
[2018-07-28 00:14:14.175][51][info][upstream] external/envoy/source/server/lds_api.cc:73] lds: add/update listener '0.0.0.0_80'
[2018-07-28 00:14:14.180][51][info][config] external/envoy/source/server/listener_manager_impl.cc:903] all dependencies initialized. starting workers
[2018-07-28 00:14:17.178][51][info][main] external/envoy/source/server/drain_manager_impl.cc:63] shutting down parent after drain
[2018-07-28 00:14:27.977][51][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:240] gRPC config stream closed: 13, 
[2018-07-28 00:19:08.103][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:394] add/update cluster outbound|80||a.istio-system.svc.cluster.local starting warming
```

Something is wrong here with envoy - it also took 5 min to reconnect.

**Comment by costinm** [comment]
+Piotr and Lizan - it seems the fix for envoy made things worse ? Or the max connection fix doesn't work.

**Comment by costinm** [comment]
It appears setting IdleTimeout results in envoy closing the connection after LDS is received, and
not connecting for 5 min.

```
[2018-07-29 03:44:46.473][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:133] cm init: all clusters initialized
[2018-07-29 03:44:46.473][51][info][main] external/envoy/source/server/server.cc:390] all clusters initialized. initializing init manager
[2018-07-29 03:44:46.553][51][info][upstream] external/envoy/source/server/lds_api.cc:73] lds: add/update listener '0.0.0.0_80'
[2018-07-29 03:44:46.561][51][info][config] external/envoy/source/server/listener_manager_impl.cc:903] all dependencies initialized. starting workers
[2018-07-29 03:44:49.558][51][info][main] external/envoy/source/server/drain_manager_impl.cc:63] shutting down parent after drain
[2018-07-29 03:45:00.711][51][warning][upstream] external/envoy/source/common/config/grpc_mux_impl.cc:240] gRPC config stream closed: 13, 
[2018-07-29 03:49:58.276][51][info][upstream] external/envoy/source/common/upstream/cluster_manager_impl.cc:394] add/update cluster outbound|80||a.istio-system.svc.cluster.local starting warming
```

Removing that line seems to get the tests to pass.

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 2]
```diff
@@ -796,6 +796,10 @@ func buildHTTPConnectionManager(env *model.Environment, node *model.Proxy, httpO
 		// Allow websocket upgrades
 		websocketUpgrade := &http_conn.HttpConnectionManager_UpgradeConfig{UpgradeType: "websocket"}
 		connectionManager.UpgradeConfigs = []*http_conn.HttpConnectionManager_UpgradeConfig{websocketUpgrade}
+		// Setting IdleTimeout to 0 seems to break most tests, causing
+		// envoy to disconnect. See PR comments.
+		// notimeout := 0 * time.Second
+		// connectionManager.IdleTimeout = &notimeout
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: But this is the crux of the fix (well for the older envoy). going to try with newer envoy (which requires setting stream_idle_timeout instead)

**Pull request merged**